### PR TITLE
[#IP-339] Migrate fp-ts version from 1.x to 2.x

### PR DESCRIPTION
**Pull request by fabriziopapi**
Update fp-ts version from 1.x to 2.x

#### List of Changes
Update fp-ts, io-ts, ts-commons and openapigen versions.
Fixed compilation errors due to different fp-ts 2.x version mode.
Fixed unit test compilation errors.
Refactor of saml.ts functions.

#### Motivation and Context
Reduce technical debt and enable dev to use new fp-ts feature.

#### How Has This Been Tested?

Already existing unit tests.

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/utils/saml.ts [hunk 1]
```diff
@@ -1118,395 +81,542 @@ export const getPreValidateResponse = (
...
+    TE.chain(stepSeven),
+    TE.chain(stepEight),
+    TE.chain(stepNine),
+    TE.chain(stepTen),
+    // check for Transform over SAML Response
+    TE.chainW(stepEleven),
+    TE.bimap(
+      error => stepTwelveLeft(error),
+      _ => stepTwelveRight(_)
     )
```
Comment (BurnedMarshal) [inline src/utils/saml.ts:620]: Great work in here. I like a lot the splitting of the giant `TE` chain. Now we can improve code clearness using semantic names for each validation step. Below an example, but you are free to use another naming convention.
The same thing must be do with the types use in input and output for all steps.
```suggestion
  return pipe(
    responseElementValidationStep,
    TE.chain(returnRequestAndResponseStep),
    TE.chain(parseSAMLRequestStep),
    TE.chain(getIssueInstantFromRequestStep),
    TE.chain(issueInstantValidationStep),
    TE.chain(assertionIssueInstantValidationStep),
    TE.chain(authnContextClassRefValidationStep),
    TE.chain(attributesValidationStep),
    TE.chain(responseIssuerValidationStep),
    TE.chain(assertionIssuerValidationStep),
    TE.chainW(transformValidationStep),
    TE.bimap(validationFailure, validationSuccess)
```

File: src/utils/saml.ts [hunk 2]
```diff
@@ -1118,395 +81,542 @@ export const getPreValidateResponse = (
...
+
+  const stepFive = (_: IStepFour): TaskEither<Error, IStepFive> =>
+    pipe(
+      TE.fromEither(
+        pipe(
+          _.RequestIssueInstant,
+          E.fromPredicate(
+            _1 => _1.getTime() <= _.IssueInstant.getTime(),
+            () =>
+              new Error("Request IssueInstant must after Request IssueInstant")
```
Comment (BurnedMarshal) [inline src/utils/saml.ts:369]:
```suggestion
              new Error("Response IssueInstant must after Request IssueInstant")
```

File: src/utils/saml.ts [hunk 3]
```diff
@@ -1118,395 +81,542 @@ export const getPreValidateResponse = (
...
-        )
-          .chain(
-            fromPredicate(
-              not(isEmptyNode),
-              () => new Error("Subject element must be not empty")
-            )
+        ),
+        E.chain(
+          E.fromPredicate(
+            not(isEmptyNode),
```
Comment (BurnedMarshal) [inline src/utils/saml.ts:411]: `not` function is deprecated, use `not` from `fp-ts/lib/Predicate` module
```suggestion
            predicate.not(isEmptyNode),
```

File: src/utils/middleware.ts [hunk 4]
```diff
@@ -1,11 +1,15 @@
...
  */
+// tslint:disable-next-line: no-submodule-imports
+import { EmailString, NonEmptyString } from "@pagopa/ts-commons/lib/strings";
 import * as express from "express";
-import { array } from "fp-ts/lib/Array";
-import { Task, task } from "fp-ts/lib/Task";
+import * as A from "fp-ts/lib/Array";
+import { pipe } from "fp-ts/lib/function";
+import * as T from "fp-ts/lib/Task";
+import { Task } from "fp-ts/lib/Task";
```
[the remark this answers is not in the document]
Comment (fabriziopapi) [reply src/utils/middleware.ts:10]: fixed, thank you!

File: src/utils/middleware.ts [hunk 5]
```diff
@@ -137,59 +141,74 @@ export const getSpidStrategyOptionsUpdater = (
...
+              TE.getOrElseW(() => T.of({}))
+            )
           ]
         : []
     );
-  return array
-    .sequence(task)(idpOptionsTasks)
-    .map(idpOptionsRecords =>
+  return pipe(
+    A.sequence(T.task)(idpOptionsTasks),
```
[the remark this answers is not in the document]
Comment (fabriziopapi) [reply src/utils/middleware.ts:198]: fixed, thank you!

File: src/utils/middleware.ts [hunk 6]
```diff
@@ -137,59 +141,74 @@ export const getSpidStrategyOptionsUpdater = (
...
+    T.map(idpOptionsRecords =>
       idpOptionsRecords.reduce((prev, current) => ({ ...prev, ...current }), {})
-    )
-    .map(idpOptionsRecord => {
+    ),
```
[the remark this answers is not in the document]
Comment (fabriziopapi) [reply src/utils/middleware.ts]: changed, thank you!

File: src/utils/saml.ts (hunk 1 above)
Comment (fabriziopapi) [reply src/utils/saml.ts]: I lazly enumerated them, now is a lot better! thank you!

File: src/utils/saml.ts (hunk 3 above)
Comment (fabriziopapi) [reply src/utils/saml.ts:411]: fixed, thank you!

**Comment by balanza** [comment]
@BurnedMarshal are we fine with latest changes?

File: src/utils/samlUtils.ts [hunk 7]
```diff
@@ -0,0 +1,1176 @@
...
+          O.fromNullable(
+            Assertion.getElementsByTagNameNS(
+              SAML_NAMESPACE.ASSERTION,
+              "Subject"
+            ).item(0)
+          )
+        ),
+        E.chain(
+          E.fromPredicate(
+            not(isEmptyNode),
```
Comment (BurnedMarshal) [inline src/utils/samlUtils.ts:785]: use `not` from `predicate`

File: src/utils/samlUtils.ts [hunk 8]
```diff
@@ -0,0 +1,1176 @@
...
+              O.fromNullable(
+                Subject.getElementsByTagNameNS(
+                  SAML_NAMESPACE.ASSERTION,
+                  "NameID"
+                ).item(0)
+              )
+            ),
+            E.chain(
+              E.fromPredicate(
+                not(isEmptyNode),
```
Comment (BurnedMarshal) [inline src/utils/samlUtils.ts:801]: ditto

File: src/utils/samlUtils.ts [hunk 9]
```diff
@@ -0,0 +1,1176 @@
...
+              O.fromNullable(
+                Assertion.getElementsByTagNameNS(
+                  SAML_NAMESPACE.ASSERTION,
+                  "Conditions"
+                ).item(0)
+              )
+            ),
+            E.chain(
+              E.fromPredicate(
+                not(isEmptyNode),
```
Comment (BurnedMarshal) [inline src/utils/samlUtils.ts:970]: ditto

File: src/utils/samlUtils.ts [hunk 10]
```diff
@@ -0,0 +1,1176 @@
...
+                  O.fromNullable(
+                    Conditions.getElementsByTagNameNS(
+                      SAML_NAMESPACE.ASSERTION,
+                      "AudienceRestriction"
+                    ).item(0)
+                  )
+                ),
+                E.chain(
+                  E.fromPredicate(
+                    not(isEmptyNode),
```
Comment (BurnedMarshal) [inline src/utils/samlUtils.ts:1017]: ditto

File: src/utils/samlUtils.ts [hunk 11]
```diff
@@ -0,0 +1,1176 @@
...
+                  O.fromNullable(
+                    Assertion.getElementsByTagNameNS(
+                      SAML_NAMESPACE.ASSERTION,
+                      "AuthnStatement"
+                    ).item(0)
+                  )
+                ),
+                E.chain(
+                  E.fromPredicate(
+                    not(isEmptyNode),
```
Comment (BurnedMarshal) [inline src/utils/samlUtils.ts:1057]: ditto

File: src/utils/samlUtils.ts [hunk 12]
```diff
@@ -0,0 +1,1176 @@
...
+                      O.fromNullable(
+                        AuthnStatement.getElementsByTagNameNS(
+                          SAML_NAMESPACE.ASSERTION,
+                          "AuthnContext"
+                        ).item(0)
+                      )
+                    ),
+                    E.chain(
+                      E.fromPredicate(
+                        not(isEmptyNode),
```
Comment (BurnedMarshal) [inline src/utils/samlUtils.ts:1073]: ditto

File: src/utils/samlUtils.ts [hunk 13]
```diff
@@ -0,0 +1,1176 @@
...
+                          O.fromNullable(
+                            AuthnContext.getElementsByTagNameNS(
+                              SAML_NAMESPACE.ASSERTION,
+                              "AuthnContextClassRef"
+                            ).item(0)
+                          )
+                        ),
+                        E.chain(
+                          E.fromPredicate(
+                            not(isEmptyNode),
```
Comment (BurnedMarshal) [inline src/utils/samlUtils.ts:1091]: ditto

File: src/utils/samlUtils.ts [hunk 14]
```diff
@@ -0,0 +1,1176 @@
...
+              O.fromNullable(
+                Subject.getElementsByTagNameNS(
+                  SAML_NAMESPACE.ASSERTION,
+                  "SubjectConfirmation"
+                ).item(0)
+              )
+            ),
+            E.chain(
+              E.fromPredicate(
+                not(isEmptyNode),
```
Comment (BurnedMarshal) [inline src/utils/samlUtils.ts:854]: ditto

File: src/utils/samlUtils.ts [hunk 15]
```diff
@@ -0,0 +1,1176 @@
...
+    .map((_, i) => signatures.item(i))
+    .some(
+      item =>
+        item?.getAttribute("Algorithm")?.valueOf() ===
+        "http://www.w3.org/2000/09/xmldsig#hmac-sha1"
+    );
+};
+
+export const notSignedWithHmacPredicate = E.fromPredicate(
+  not(isSignedWithHmac),
```
Comment (BurnedMarshal) [inline src/utils/samlUtils.ts:93]: ditto

**Review by BurnedMarshal — CHANGES_REQUESTED** [review]

File: src/utils/samlUtils.ts [hunk 16]
```diff
@@ -0,0 +1,1176 @@
+/**
```
[the remark this answers is not in the document]
Comment (fabriziopapi) [reply src/utils/samlUtils.ts:1]: 'saml.ts' has not be renamed: I created 'samlUtils.ts' to contains part of 'saml.ts' functions. How could I use 'git mv' on saml.ts to not loose saml.ts itself?

File: src/utils/samlUtils.ts [hunk 17]
```diff
@@ -0,0 +1,1176 @@
...
+ */
+// tslint:disable-next-line: no-submodule-imports
+import { UTCISODateFromString } from "@pagopa/ts-commons/lib/dates";
+// tslint:disable-next-line: no-submodule-imports
+import { NonEmptyString } from "@pagopa/ts-commons/lib/strings";
+import { distanceInWordsToNow, isAfter, subDays } from "date-fns";
+import { Request as ExpressRequest } from "express";
+import { flatten } from "fp-ts/lib/Array";
+import * as E from "fp-ts/lib/Either";
+import { Either } from "fp-ts/lib/Either";
```
[the remark this answers is not in the document]
Comment (fabriziopapi) [reply src/utils/samlUtils.ts:15]: fixed. thank you!

File: src/utils/samlUtils.ts [hunk 18]
```diff
@@ -0,0 +1,1176 @@
...
+import { not, pipe } from "fp-ts/lib/function";
+import * as O from "fp-ts/lib/Option";
+import { Option } from "fp-ts/lib/Option";
```
[the remark this answers is not in the document]
Comment (fabriziopapi) [reply src/utils/samlUtils.ts:18]: fixed. thank you!

File: src/utils/samlUtils.ts [hunk 19]
```diff
@@ -0,0 +1,1176 @@
...
+import { collect, lookup } from "fp-ts/lib/Record";
+import { Ord } from "fp-ts/lib/string";
+import * as TE from "fp-ts/lib/TaskEither";
+import { TaskEither } from "fp-ts/lib/TaskEither";
```
[the remark this answers is not in the document]
Comment (fabriziopapi) [reply src/utils/samlUtils.ts:22]: fixed. thank you!

File: src/utils/samlUtils.ts (hunk 15 above)
Comment (fabriziopapi) [reply src/utils/samlUtils.ts:93]: fixed. thank you!

File: src/utils/samlUtils.ts (hunk 7 above)
Comment (fabriziopapi) [reply src/utils/samlUtils.ts:785]: fixed. thank you!

File: src/utils/samlUtils.ts (hunk 8 above)
Comment (fabriziopapi) [reply src/utils/samlUtils.ts:801]: fixed. thank you!

File: src/utils/samlUtils.ts (hunk 14 above)
Comment (fabriziopapi) [reply src/utils/samlUtils.ts:854]: fixed. thank you!

File: src/utils/samlUtils.ts (hunk 9 above)
Comment (fabriziopapi) [reply src/utils/samlUtils.ts:970]: fixed!

File: src/utils/samlUtils.ts (hunk 10 above)
Comment (fabriziopapi) [reply src/utils/samlUtils.ts:1017]: fixed!

File: src/utils/samlUtils.ts (hunk 11 above)
Comment (fabriziopapi) [reply src/utils/samlUtils.ts:1057]: fixed!

File: src/utils/samlUtils.ts (hunk 12 above)
Comment (fabriziopapi) [reply src/utils/samlUtils.ts:1073]: fixed!

File: src/utils/samlUtils.ts (hunk 13 above)
Comment (fabriziopapi) [reply src/utils/samlUtils.ts:1091]: fixed!

File: src/utils/samlUtils.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (BurnedMarshal) [reply src/utils/samlUtils.ts:1]: You are right, sorry my bad. IMHO to better see changes we should move functions to others files after the refactoring. `git mv` make sense only if we move/rename the entire file.

[the remark this answers is not in the document]
Comment (fabriziopapi) [reply src/utils/samlUtils.ts:1]: Git is a lot tricky on this point. I tried to rework the branch to keep the changes: if you go on the last commit 'update saml.ts (with history)' you should see the changes. 
(https://github.com/pagopa/io-spid-commons/pull/91/commits/7e83fd96ea4053725a6974a64279474d3c64ea36)

File: src/utils/samlUtils.ts [hunk 20]
```diff
@@ -0,0 +1,1176 @@
...
+                                SPID_LEVELS.SpidL2 ||
+                              AuthnContextClassRef.textContent?.trim() ===
+                                SPID_LEVELS.SpidL3,
+                            () =>
+                              new Error("Invalid AuthnContextClassRef value")
+                          )
+                        ),
+                        E.map(AuthnContextClassRef =>
+                          AuthnContextClassRef.textContent?.trim()
+                        ),
```
Comment (BurnedMarshal) [inline src/utils/samlUtils.ts:1110]: We can optimize this code doing the map before the chain
```suggestion
                        E.map(AuthnContextClassRef =>
                          AuthnContextClassRef.textContent?.trim()
                        ),
                        E.chain(
                          E.fromPredicate(
                            AuthnContextClassRef =>
                              AuthnContextClassRef === SPID_LEVELS.SpidL1 ||
                              AuthnContextClassRef === SPID_LEVELS.SpidL2 ||
                              AuthnContextClassRef === SPID_LEVELS.SpidL3,
                            () =>
                              new Error("Invalid AuthnContextClassRef value")
                          )
                        ),
```

File: src/utils/saml.ts [hunk 21]
```diff
@@ -1118,395 +82,557 @@ export const getPreValidateResponse = (
...
+  type IStepThree = IStepTwo & { Request: Document };
+
+  type IStepFour = IStepThree & {
+    RequestIssueInstant: Date;
+    RequestAuthnRequest: Element;
+  };
+
+  type IStepFive = IStepFour;
+
+  type IStepSix = IStepFive;
```
Comment (BurnedMarshal) [inline src/utils/saml.ts:123]: Remove Alias type

File: src/utils/saml.ts [hunk 22]
```diff
@@ -1118,395 +82,557 @@ export const getPreValidateResponse = (
...
+
+  type IStepSeven = IStepSix & { RequestAuthnContextClassRef: NonEmptyString };
+
+  type IStepEight = IStepSeven;
+
+  type IStepNine = IStepEight;
+
+  type IStepTen = IStepNine;
```
Comment (BurnedMarshal) [inline src/utils/saml.ts:131]: ditto

File: src/utils/saml.ts [hunk 23]
```diff
@@ -1118,395 +82,557 @@ export const getPreValidateResponse = (
...
+
+  interface ITransformValidation {
+    idpIssuer: string;
+    message: string;
+    numberOfTransforms: number;
+  }
+
+  type IStepEleven = IStepTen;
```
Comment (BurnedMarshal) [inline src/utils/saml.ts:139]: ditto

File: src/utils/saml.ts [hunk 24]
```diff
@@ -1118,395 +82,557 @@ export const getPreValidateResponse = (
...
+    AssertionIssueInstant: Date;
+  }
+
+  interface ISamlCacheType {
+    RequestXML: string;
+    createdAt: Date;
+    idpIssuer: string;
+  }
+
+  type IStepTwo = IBaseOutput & {
```
Comment (BurnedMarshal) [inline src/utils/saml.ts:110]: Don't use generic `StepX` types names, use instead semantic types names to clearly identify what these types are.

**Review by BurnedMarshal — APPROVED** [review]
LGTM apart latest minor changes. Thank for the huge work in this PR.

File: src/utils/samlUtils.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (BurnedMarshal) [reply src/utils/samlUtils.ts:1]: Thanks, it was a lot helpful

File: src/utils/saml.ts (hunk 21 above)
Comment (fabriziopapi) [reply src/utils/saml.ts:123]: you are right! fixed!

File: src/utils/saml.ts (hunk 22 above)
Comment (fabriziopapi) [reply src/utils/saml.ts]: fixed!

File: src/utils/saml.ts (hunk 23 above)
Comment (fabriziopapi) [reply src/utils/saml.ts]: fixed!

File: src/utils/saml.ts (hunk 24 above)
Comment (fabriziopapi) [reply src/utils/saml.ts]: fixed!

File: src/utils/samlUtils.ts (hunk 20 above)
Comment (fabriziopapi) [reply src/utils/samlUtils.ts]: nice catch! fixed!

**Pull request merged**